dace/codegen/targets/fpga.py

```diff
@@ -421,10 +421,11 @@ def find_rtl_tasklet(self, subgraph: ScopeSubgraphView):
         '''
         for n in subgraph.nodes():
             if isinstance(n, dace.nodes.NestedSDFG):
-                for sg in dace.sdfg.concurrent_subgraphs(n.sdfg.start_state):
-                    node = self.find_rtl_tasklet(sg)
-                    if node:
-                        return node
+                if len(n.sdfg.nodes()) == 1 and isinstance(n.sdfg.nodes()[0], SDFGState):
```

Comment:
Yes there were actually many cases, in which the nested SDFG would (on main, so DaCe 1.0) consist of a state machine to represent a loop or something similar. This function (I assume - but I do not know the FPGA backend well enough: probably incorrectly) only examined concurrent subgraphs in the first state of that state machine. That would typically be an empty list, since those states are very often empty (meta states for loop initialization, for instance). After this change to DaCe 2.0, anything with > 1 state would often be replaced with a control flow region, thus exposing the issue. This change prevents this from causing issues, first making sure that the nested SDFG contains only a single state, as the author of the function probably intended. If the function is supposed to check all states of a nested SDFG, it is entirely wrong to begin with.

A 'log' of what went wrong can be seen in this specific test run, for instance: https://github.com/spcl/dace/actions/runs/11485319192/job/31965101720
Running any of these tests (for instance `overapprox_transient_shapes_test.py`) on current main and placing a breakpoint in these two methods that contain the change here, quickly exposes the issue.

I thus assume that this is an underlying bug that needs to be fixed - however, again, I do not understand the FPGA backend well enough to attempt such a fix.

Reply:
This is good documentation of the codegen issue. Please add a comment on the codegen refactor project referring to this line.
EDIT: Now that the codegen refactor project is gone for some reason (it was a legacy project), we should create a new one.

Reply:
Tracking in https://github.com/orgs/spcl/projects/16